File: aplicaciones_demostrativas/JAVA/JAVAmwEIDTest_NetBeans-6.8/src/javamweidtest/PKCS11Controller.java
package javamweidtest;

import iaik.pkcs.pkcs11.*;
import iaik.pkcs.pkcs11.objects.*;
import iaik.pkcs.pkcs11.wrapper.PKCS11Constants;

import java.io.*;
import java.util.*;

class EncryptedData {
    private byte[] m_EncryptedData;
    public void SetBytes(byte[] in_bytes)
    {
        m_EncryptedData = in_bytes;
    }
    public String ToString()
    {
        if(m_EncryptedData == null) {
            return "";
        }
        return new String(m_EncryptedData);
    }
}


public class PKCS11Controller {

    private String m_FileName;
    private String m_AutenticacionLabel = "NOT_SET";
    private String m_SignLabel = "NOT_SET";
    private String m_IssuerCertificate = "NOT_SET";
    private Module m_Module = null;
    private Slot[] m_Slots = null;
    private int m_CurrentIndex = -1;

    public PKCS11Controller(String in_FileName) {
        m_FileName = in_FileName;
    }

    public List<String> GetReadersList(boolean in_Refresh) {
        //List<String> readers = new LinkedList<>(); //jdk7
    	List<String> readers = new LinkedList<String>(); //jdk6
        try {
            if (m_Module == null) {
                m_Module = Module.getInstance(m_FileName);
                m_Module.initialize(null);
            }

            if (m_Slots == null) {
                // GetSlotList.
                m_Slots = m_Module.getSlotList(true);
            } else if (in_Refresh) {
                //TODO verify
                //m_Slots = m_Module.GetSlotList(true);
            }
            for (Slot slot : m_Slots) {
                readers.add(slot.getSlotInfo().getSlotDescription());
                System.out.println(slot.getSlotInfo().getSlotDescription());
            }
        } 
        catch (IOException e) {
            System.out.println(e);
        }
        catch (TokenException e) {
            System.out.println(e);
        }
        return readers;
    }

    public boolean NumeroTarjetaValida(int in_SlotIndex, String in_NumeroTarjeta) {
        boolean result = false;
        if (m_Slots.length > in_SlotIndex) {
            try {
                Slot slot = m_Slots[in_SlotIndex];
                result = slot.getToken().getTokenInfo().getSerialNumber().matches(in_NumeroTarjeta);
            } catch (TokenException ex) {
                System.out.println(ex);
            }
        }
        return result;
    }

    public boolean Login(int in_SlotIndex, String in_PIN) {
        boolean result = false;
        try {
            if (m_Module == null) {
                m_Module = Module.getInstance(m_FileName);
                m_Module.initialize(null);
            }

            if (m_Slots == null) {
                // GetSlotList.
                m_Slots = m_Module.getSlotList(true);
            }
            if (m_Slots.length > in_SlotIndex) {
                Slot slot = m_Slots[in_SlotIndex];
                Session session = slot.getToken().openSession(Token.SessionType.SERIAL_SESSION,
                        Token.SessionReadWriteBehavior.RO_SESSION, null, null);
                m_CurrentIndex = in_SlotIndex;
                session.login(Session.UserType.USER, in_PIN.toCharArray());

                try {

                    GenericTemplate certificateSearchTemplate = new GenericTemplate();
                    LongAttribute objectClassAttribute = new LongAttribute(PKCS11Constants.CKA_CLASS);
                    objectClassAttribute.setLongValue(new Long(PKCS11Constants.CKO_CERTIFICATE));
                    certificateSearchTemplate.addAttribute(objectClassAttribute);
                    LongAttribute certificateTypeAttribute = new LongAttribute(PKCS11Constants.CKA_CERTIFICATE_TYPE);
                    certificateTypeAttribute.setLongValue(new Long(PKCS11Constants.CKC_X_509));
                    certificateSearchTemplate.addAttribute(certificateTypeAttribute);


                    session.findObjectsInit(certificateSearchTemplate);
                    //P11Object
                    iaik.pkcs.pkcs11.objects.Object[] certificates = session.findObjects(2);
                    if (certificates.length == 2) {
                        SetAutenticacionLabel(new String(((X509PublicKeyCertificate) certificates[0]).getLabel().getCharArrayValue()));
                        SetSignatureLabel(new String(((X509PublicKeyCertificate) certificates[1]).getLabel().getCharArrayValue()));
                    }

                    session.findObjectsFinal();
                    ///////////////////
                    result = true;
                } finally {
                    // Log out.
                    session.logout();
                }
            }
        } 
        catch (IOException e) {
            System.out.println(e);
        }
        catch (TokenException e) {
            System.out.println(e);
        }
        return result;
    }

    public boolean Firmar(int in_SlotIndex, String in_PIN, byte[] in_Data, EncryptedData out_encryptedData) {
        boolean result = false;
        out_encryptedData.SetBytes(null);
        try {
            if (m_Module == null) {
                m_Module = Module.getInstance(m_FileName);
                m_Module.initialize(null);
            }

            if (m_Slots == null) {
                // GetSlotList.
                m_Slots = m_Module.getSlotList(true);
            }
            if (m_Slots.length > in_SlotIndex) {
                Slot slot = m_Slots[in_SlotIndex];
                Session session = slot.getToken().openSession(Token.SessionType.SERIAL_SESSION,
                        Token.SessionReadWriteBehavior.RO_SESSION, null, null);
                m_CurrentIndex = in_SlotIndex;
                session.login(Session.UserType.USER, in_PIN.toCharArray());

                try {

                    GenericTemplate privateKeySearchTemplate = new GenericTemplate();
                    LongAttribute classAttribute = new LongAttribute(PKCS11Constants.CKA_CLASS);
                    classAttribute.setLongValue(new Long(PKCS11Constants.CKO_PRIVATE_KEY));
                    privateKeySearchTemplate.addAttribute(classAttribute);

                    ByteArrayAttribute keyLabelAttribute = new ByteArrayAttribute(PKCS11Constants.CKA_LABEL);
                    
                    byte[] label = m_SignLabel.getBytes("UTF-8");
                    
                    keyLabelAttribute.setByteArrayValue(label);
                    privateKeySearchTemplate.addAttribute(keyLabelAttribute);

                    session.findObjectsInit(privateKeySearchTemplate);

                    iaik.pkcs.pkcs11.objects.Object[] privatekeys = session.findObjects(1);

                    session.findObjectsFinal();

                    if (privatekeys.length >= 1) {
                        
                        iaik.pkcs.pkcs11.Mechanism signatureMechanism = iaik.pkcs.pkcs11.Mechanism.get(PKCS11Constants.CKM_RSA_PKCS);
                        // initialize for signing
                        session.signInit(signatureMechanism, (Key) privatekeys[0]);                        
                        
                        out_encryptedData.SetBytes(session.sign(in_Data));
                    }
                    result = true;
                } finally {
                    // Log out.
                    session.logout();
                }
            }
        } 
        catch (IOException e) {
            System.out.println(e);
        }
        catch (TokenException e) {
            System.out.println(e);
        }
        return result;

    }

    void SetAutenticacionLabel(String in_AuthenticationLabel) {
        m_AutenticacionLabel = in_AuthenticationLabel;
    }

    void SetSignatureLabel(String in_SignLabel) {
        m_SignLabel = in_SignLabel;
    }

    void SetIssuerCertificate(String in_IssuerCertificate) {
        m_IssuerCertificate = in_IssuerCertificate;
    }

    String GetTokenInfo(int in_SlotIndex) {
        if (m_Slots.length > in_SlotIndex) {
            try {
                Slot slot = m_Slots[in_SlotIndex];
                StringBuilder builder = new StringBuilder();
                builder.append("ManufacturerID: ");
                builder.append(slot.getToken().getTokenInfo().getManufacturerID());
                builder.append("\n");
                builder.append("Label: ");
                builder.append(slot.getToken().getTokenInfo().getLabel());
                builder.append("\n");
                builder.append("HardwareVersion: ");
                builder.append(slot.getToken().getTokenInfo().getHardwareVersion().toString());
                builder.append("\n");
                builder.append("FirmwareVersion: ");
                builder.append(slot.getToken().getTokenInfo().getFirmwareVersion().toString());
                builder.append("\n");
                builder.append("Model: ");
                builder.append(slot.getToken().getTokenInfo().getModel());
                return builder.toString();
            } catch (TokenException ex) {
                System.out.println(ex);
            }
        }
        return "";
    }

    boolean Autenticar(int in_SlotIndex, String in_PIN, StringBuilder out_Error) {
        boolean result = false;

        try {
            if (m_Module == null) {
                m_Module = Module.getInstance(m_FileName);
                m_Module.initialize(null);
            }

            if (m_Slots == null) {
                // GetSlotList.
                m_Slots = m_Module.getSlotList(true);
            }
            if (m_Slots.length > in_SlotIndex) {
                Slot slot = m_Slots[in_SlotIndex];
                Session session = slot.getToken().openSession(Token.SessionType.SERIAL_SESSION,
                        Token.SessionReadWriteBehavior.RO_SESSION, null, null);
                m_CurrentIndex = in_SlotIndex;
                session.login(Session.UserType.USER, in_PIN.toCharArray());

                try {
                    
                    GenericTemplate certificateSearchTemplate = new GenericTemplate();
                    LongAttribute classAttribute = new LongAttribute(PKCS11Constants.CKA_CLASS);
                    classAttribute.setLongValue(new Long(PKCS11Constants.CKO_CERTIFICATE));
                    certificateSearchTemplate.addAttribute(classAttribute);
                    
                    LongAttribute certificateTypeAttribute = new LongAttribute(PKCS11Constants.CKA_CERTIFICATE_TYPE);
                    certificateTypeAttribute.setLongValue(new Long(PKCS11Constants.CKC_X_509));
                    certificateSearchTemplate.addAttribute(certificateTypeAttribute);
                    
                    ByteArrayAttribute keyLabelAttribute = new ByteArrayAttribute(PKCS11Constants.CKA_LABEL);
                    
                    byte[] label = m_AutenticacionLabel.getBytes("UTF-8");
                    
                    keyLabelAttribute.setByteArrayValue(label);
                    certificateSearchTemplate.addAttribute(keyLabelAttribute);

                    session.findObjectsInit(certificateSearchTemplate);

                    iaik.pkcs.pkcs11.objects.Object[] certificates = session.findObjects(1);

                    session.findObjectsFinal();

                    if (certificates.length >= 1) {
                        
                        X509PublicKeyCertificate cert = (X509PublicKeyCertificate) certificates[0];
                        OcspClient oscpClient = new OcspClient(cert.getValue().getByteArrayValue());
                        try {
                            oscpClient.getPublicKeyCertificate().checkValidity();
                            CertificateStatus status = oscpClient.consultarEstadoDeCertificado(oscpClient.getPublicKeyCertificate(), oscpClient.leerCertificado(m_IssuerCertificate));
                            if (status == CertificateStatus.Good) {
                                out_Error.append("OK");
                                result = true;
                            } else if (status == CertificateStatus.Revoked) {
                                out_Error.append("Certificado Revocado");
                            } else {
                                out_Error.append("Certificado/Respuesta Desconocido(s)");
                            }
                        }
                        catch(Exception ex) {
                            System.out.println(ex);
                            out_Error.append( ex.getMessage());
                        }
                    } else {
                        out_Error.append( "No se encontraron objetos en la tarjeta.");
                    }


                } catch (UnsupportedEncodingException  e) {
                    System.out.println(e);
                    out_Error.append( e.getMessage());
                }
                catch (TokenException e) {
                    System.out.println(e);
                    out_Error.append( e.getMessage());
                } 
                finally {
                    // Log out.
                    session.logout();
                }
            }
        } 
        catch (IOException e) {
            System.out.println(e);
        }
        catch (TokenException e) {
            System.out.println(e);
        }
        return result;
    }
}

File: aplicaciones_demostrativas/mw/CS/CSmwEIDTest_VisualStudio-2010/CSmwEIDTest/PKCS11Controller.cs
using System;
using System.Collections.Generic;

using System.Text;

using System.Runtime.InteropServices;

using Net.Sf.Pkcs11;
using Net.Sf.Pkcs11.Objects;
using Net.Sf.Pkcs11.Wrapper;

using System.Security.Cryptography.X509Certificates;
using System.IO;


namespace CSmwEIDTest
{
    class PKCS11Controller
    {
        private String m_FileName;
        private String m_AutenticacionLabel = "NOT_SET";
        private String m_SignLabel = "NOT_SET";
        private String m_IssuerCertificate = "NOT_SET";

        private Module m_Module = null;
        private Slot[] m_Slots = null;
        private int m_CurrentIndex = -1;
        public PKCS11Controller(String in_FileName)
        {
            m_FileName = in_FileName;
        }
        public List<String> GetReadersList(bool in_Refresh = false)
        {
            List<String> readers = new List<string>();
            try
            {
                if (m_Module == null)
                {
                    m_Module = Module.GetInstance(m_FileName);
                }

                if (m_Slots == null)
                {
                    // GetSlotList.
                    m_Slots = m_Module.GetSlotList(true);
                }
                else if (in_Refresh)
                {
                    //TODO verify
                    //m_Slots = m_Module.GetSlotList(true);
                }
                foreach (Slot slot in m_Slots)
                {
                    readers.Add(slot.SlotInfo.SlotDescription);
                    Console.WriteLine(slot.SlotInfo.SlotDescription);
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
            }
            return readers;
        }

        internal bool NumeroTarjetaValida(int in_SlotIndex, string in_NumeroTarjeta)
        {
            bool result = false;
            if (m_Slots.Length > in_SlotIndex)
            {
                Slot slot = m_Slots[in_SlotIndex];
                result = slot.Token.TokenInfo.SerialNumber == in_NumeroTarjeta;
            }
            return result;
        }

        internal bool Login(int in_SlotIndex, string in_PIN)
        {
            bool result = false;
            try
            {
                if (m_Module == null)
                {
                    m_Module = Module.GetInstance(m_FileName);
                }

                if (m_Slots == null)
                {
                    // GetSlotList.
                    m_Slots = m_Module.GetSlotList(true);
                }
                if (m_Slots.Length > in_SlotIndex)
                {
                    Slot slot = m_Slots[in_SlotIndex];
                    Session session = slot.Token.OpenSession(false);
                    m_CurrentIndex = in_SlotIndex;
                    session.Login(UserType.USER, in_PIN);

                    try
                    {

                        ObjectClassAttribute classAttribute = new ObjectClassAttribute(CKO.CERTIFICATE);
                        //ByteArrayAttribute keyLabelAttribute = new ByteArrayAttribute(CKA.LABEL);
                        //keyLabelAttribute.Value = System.Text.Encoding.UTF8.GetBytes(m_SignLabel);

                        session.FindObjectsInit(new P11Attribute[] {
                                 classAttribute
                          //       keyLabelAttribute
                                }
                                );
                        P11Object[] certificates = session.FindObjects(2) as P11Object[];
                        if (certificates.Length == 2)
                        {
                            SetAutenticacionLabel(new string(((X509PublicKeyCertificate)certificates[0]).Label.Value));
                            SetSignatureLabel(new string(((X509PublicKeyCertificate)certificates[1]).Label.Value));
                        }
                        
                        session.FindObjectsFinal();

                        ///////////////////
                        result = true;
                    }
                    finally
                    {
                        // Log out.
                        session.Logout();
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
            }
            return result;
        }



        internal bool Firmar(int in_SlotIndex, string in_PIN, byte[] in_Data, out byte[] out_encryptedData)
        {
            bool result = false;
            out_encryptedData = null;
            try
            {
                if (m_Module == null)
                {
                    m_Module = Module.GetInstance(m_FileName);
                }

                if (m_Slots == null)
                {
                    // GetSlotList.
                    m_Slots = m_Module.GetSlotList(true);
                }
                if (m_Slots.Length > in_SlotIndex)
                {
                    Slot slot = m_Slots[in_SlotIndex];
                    Session session = slot.Token.OpenSession(false);
                    m_CurrentIndex = in_SlotIndex;
                    session.Login(UserType.USER, in_PIN);

                    try
                    {
                        ObjectClassAttribute classAttribute = new ObjectClassAttribute(CKO.PRIVATE_KEY);
                        ByteArrayAttribute keyLabelAttribute = new ByteArrayAttribute(CKA.LABEL);
                        keyLabelAttribute.Value = System.Text.Encoding.UTF8.GetBytes(m_SignLabel);

                        session.FindObjectsInit(new P11Attribute[] {
                                 classAttribute,
                                 keyLabelAttribute
                                }
                                );
                        P11Object[] privatekeys = session.FindObjects(1) as P11Object[];
                        session.FindObjectsFinal();

                        if (privatekeys.Length >= 1)
                        {
                            session.SignInit(new Mechanism(CKM.SHA1_RSA_PKCS), (PrivateKey)privatekeys[0]);
                            out_encryptedData = session.Sign(in_Data);
                        }
                        result = true;
                    }
                    finally
                    {
                        // Log out.
                        session.Logout();
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
            }
            return result;

        }

        internal void SetAutenticacionLabel(string in_AuthenticationLabel)
        {
            m_AutenticacionLabel = in_AuthenticationLabel;
        }

        internal void SetSignatureLabel(string in_SignLabel)
        {
            m_SignLabel = in_SignLabel;
        }
        internal void SetIssuerCertificate(string in_IssuerCertificate)
        {
            m_IssuerCertificate = in_IssuerCertificate;
        }
        
        internal string GetTokenInfo(int in_SlotIndex)
        {
            if (m_Slots.Length > in_SlotIndex)
            {
                Slot slot = m_Slots[in_SlotIndex];
                StringBuilder builder = new StringBuilder();
                builder.Append("ManufacturerID: ");
                builder.Append(slot.Token.TokenInfo.ManufacturerID);
                builder.AppendLine();
                builder.Append("Label: ");
                builder.Append(slot.Token.TokenInfo.Label);
                builder.AppendLine();
                builder.Append("HardwareVersion: ");
                builder.Append(slot.Token.TokenInfo.HardwareVersion.ToString());
                builder.AppendLine();
                builder.Append("FirmwareVersion: ");
                builder.Append(slot.Token.TokenInfo.FirmwareVersion.ToString());
                builder.AppendLine();
                builder.Append("Model: ");
                builder.Append(slot.Token.TokenInfo.Model);
                return builder.ToString();
            }
            return String.Empty;
        }

        internal bool Autenticar(int in_SlotIndex, string in_PIN, out string out_Error)
        {
            bool result = false;
            out_Error = "OK";

            try
            {
                if (m_Module == null)
                {
                    m_Module = Module.GetInstance(m_FileName);
                }

                if (m_Slots == null)
                {
                    // GetSlotList.
                    m_Slots = m_Module.GetSlotList(true);
                }
                if (m_Slots.Length > in_SlotIndex)
                {
                    Slot slot = m_Slots[in_SlotIndex];
                    Session session = slot.Token.OpenSession(false);
                    m_CurrentIndex = in_SlotIndex;
                    session.Login(UserType.USER, in_PIN);

                    try
                    {
                        ObjectClassAttribute certificateAttribute = new ObjectClassAttribute(CKO.CERTIFICATE);
                        ByteArrayAttribute fileLabel = new ByteArrayAttribute(CKA.LABEL);
                        fileLabel.Value = System.Text.Encoding.UTF8.GetBytes(m_AutenticacionLabel);

                        session.FindObjectsInit(new P11Attribute[] {
                                 certificateAttribute,
                                 fileLabel
                                }
                                );
                        P11Object[] foundObjects = session.FindObjects(1) as P11Object[];

                        if (foundObjects.Length == 1)
                        {
                            X509PublicKeyCertificate cert = foundObjects[0] as X509PublicKeyCertificate;
                            OcspClient oscpClient = new OcspClient(cert.Value.Encode());
                            if (oscpClient.PublicKeyCertificate.IsValidNow)
                            {
                                CertificateStatus status = oscpClient.ConsultarEstadoDeCertificado(oscpClient.PublicKeyCertificate, oscpClient.LeerCertificado(m_IssuerCertificate));
                                if (status == CertificateStatus.Good)
                                {
                                    result = true;
                                }
                                else if (status == CertificateStatus.Revoked)
                                {
                                    out_Error = "Certificado Revocado";
                                }
                                else
                                {
                                    out_Error = "Certificado Desconocido";
                                }
                            }
                            else
                            {
                                out_Error = "Certificado Expirado";
                            }
                        }
                        else
                        {
                            out_Error = "No se encontraron objetos en la tarjeta.";
                        }

                        session.FindObjectsFinal();
                        
                    }
                    catch( System.Net.WebException wex) 
                    {
                        Console.WriteLine(wex.ToString());
                        out_Error = wex.Message;
                    }
                    catch (Exception e)
                    {
                        Console.WriteLine(e.ToString());
                        out_Error = e.Message;
                    }
                    finally
                    {
                        // Log out.
                        session.Logout();
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
            }
            return result;
        }
        
    }
}
